kedro/extras/datasets/hdf5/h5py_dataset.py

```diff
@@ -0,0 +1,217 @@
+"""``H5pyDataSet`` loads/saves data from/to a hdf file using an underlying
+filesystem (e.g.: local, S3, GCS). It uses h5py.File to handle the hdf file.
+"""
+
+import tempfile
+from copy import deepcopy
+from io import BytesIO
+from pathlib import PurePosixPath
+from threading import Lock
+from typing import Any, Dict
+
+import fsspec
+import h5py
+
+from kedro.io.core import (
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+
+class H5pyDataSet(AbstractVersionedDataSet[h5py.File, h5py.File]):
+    """``H5pyDataSet`` loads/saves data from/to a hdf file using an underlying
+    filesystem (e.g. local, S3, GCS). It uses h5py.File to handle the hdf file.
+
+    Example adding a catalog entry with
+    `YAML API <https://kedro.readthedocs.io/en/stable/data/\
+        data_catalog.html#use-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> hdf_dataset:
+        >>>   type: hdf5.H5pyDataSet
+        >>>   filepath: s3://my_bucket/raw/sensor_reading.h5
+        >>>   credentials: aws_s3_creds
+
+    Example using Python API:
+    ::
+
+        >>> from kedro.extras.datasets.hdf5 import H5pyDataSet
+        >>> from tempfile import TemporaryFile
+        >>> import numpy as np
+        >>> import h5py
+        >>>
+        >>> file = TemporaryFile()
+        >>> h5f = h5py.File(file)
+        >>> dataset = h5f.create_dataset('foo', data=[1,2])
+        >>>
+        >>> # data_set = H5pyDataSet(filepath="gcs://bucket/test.hdf")
+        >>> data_set = H5pyDataSet(filepath="test.h5")
+        >>> # Saved and loaded objects are both h5py.File instances
+        >>> data_set.save(h5f)
+        >>> reloaded = data_set.load()
+        >>> assert 'foo' in reloaded
+        >>> np.testing.assert_equal(reloaded['foo'][()], [1, 2])
+        >>>
+        >>> h5f.close()
+        >>> file.close()
+
+    """
+
+    # _lock is a class attribute that will be shared across all the instances.
+    # It is used to make dataset safe for threads.
+    _lock = Lock()
+    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
+    DEFAULT_SAVE_ARGS = {}  # type: Dict[str, Any]
+
+    # pylint: disable=protected-access
+    @staticmethod
+    def __h5py_from_binary(binary, load_args):
```

Comment:
Can these be just `_` instead of `__`?  also would be good to add the type for `binary`